services/scheduler/scheduler.go

```diff
@@ -0,0 +1,327 @@
+		}
+
+		_ = s.scheduler.RemoveByTag(id)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reload removes job from scheduler and add it again from DB.
+func (s *Service) Reload(id string) error {
+	dbTask, err := models.FindScheduledTaskByID(s.db.Querier, id)
+	if err != nil {
+		return err
+	}
+
+	if dbTask.Running {
+		return fmt.Errorf("task is running")
+	}
+
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+
+	task, err := s.convertDBTask(dbTask)
+	if err != nil {
+		return err
+	}
+
+	_ = s.scheduler.RemoveByTag(id)
+
+	// Don't add it to scheduler, if it's disabled.
+	if dbTask.Disabled {
+		return nil
+	}
+
+	j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+	if !dbTask.StartAt.IsZero() {
+		j = j.StartAt(dbTask.StartAt)
+	}
+
+	fn := s.wrapTask(task, dbTask.ID)
+	if _, err := j.Tag(dbTask.ID).Do(fn); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) loadFromDB() error {
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+
+	disabled := false
+	dbTasks, err := models.FindScheduledTasks(s.db.Querier, models.ScheduledTasksFilter{
+		Disabled: &disabled,
+	})
+	if err != nil {
+		return err
+	}
+
+	tasks := make([]Task, 0, len(dbTasks))
+	for _, dbTask := range dbTasks {
+		task, err := s.convertDBTask(dbTask)
+		if err != nil {
+			return err
+		}
+		tasks = append(tasks, task)
+	}
+
+	s.scheduler.Clear()
+	for i, task := range tasks {
+		dbTask := dbTasks[i]
+		fn := s.wrapTask(task, dbTask.ID)
+		j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+		if !dbTask.StartAt.IsZero() {
+			j = j.StartAt(dbTask.StartAt)
+		}
+		if _, err := j.Tag(dbTask.ID).Do(fn); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+func (s *Service) wrapTask(task Task, id string) func() {
+	return func() {
+		var err error
+		l := s.l.WithFields(logrus.Fields{
+			"id":       id,
+			"taskType": task.Type(),
+		})
+		ctx, cancel := context.WithCancel(context.Background())
+
+		s.taskMx.Lock()
+		s.tasks[id] = cancel
+		s.taskMx.Unlock()
+
+		defer func() {
+			cancel()
+			s.taskMx.Lock()
+			delete(s.tasks, id)
+			s.taskMx.Unlock()
+		}()
+
+		t := time.Now()
+		l.Debug("Starting task")
+		_, err = models.ChangeScheduledTask(s.db.Querier, id, models.ChangeScheduledTaskParams{
+			Running: pointer.ToBool(true),
+		})
+
+		if err != nil {
+			l.Errorf("failed to change running state: %v", err)
+		}
+
+		taskErr := task.Run(ctx)
+		if taskErr != nil {
+			l.Error(taskErr)
+		}
+		l.WithField("duration", time.Since(t)).Debug("Ended task")
+
+		_, err = models.ChangeScheduledTask(s.db.Querier, id, models.ChangeScheduledTaskParams{
+			Running: pointer.ToBool(false),
+		})
+
+		if err != nil {
+			l.Errorf("failed to change running status: %v", err)
+		}
+
+		s.taskFinished(id, taskErr)
+	}
+}
+
+func (s *Service) taskFinished(id string, taskErr error) {
+	var job *gocron.Job
+	for _, j := range s.scheduler.Jobs() {
+		if len(j.Tags()) > 0 && j.Tags()[0] == id {
+			job = j
+			break
+		}
+	}
```

Comment:
let's use some map to store taskID -> Job link

Reply:
Then it's another abstraction over library jobs. And internal if I do e.g remove job by tag. It works exactly the same way as here.

But okay. :+1:

Reply:
If you will have it, you will be able to just update job instead of removing and readding it.

Reply:
There is no API for updating job.

Reply:
https://github.com/go-co-op/gocron/blob/master/scheduler.go#L900

Reply:
I looked on to it. And I can't update function(+argument, even though we don't pass any). So if something changed inside. For example in PrintTask message. We can't reflect this change.

Reply:
Why would you update function? If we generate PrintTask inside function based on task id, then everything will work. We just should update cron.